[Refactor] Redis를 활용해 전송된 인증코드 유효기간을 설정한다.

ghrltjdtprbs opened the pull request:
# ⭐️ [Refactor] Redis를 활용해 인증코드 만료시간을 설정한다. 

## 변경 사항

- 인증 코드 유효기간 설정을 위해 ConcurrentHashMap을 사용하는 대신 Redis를 도입하였습니다. 
Redis의 만료 기능을 활용하여 인증 코드의 자동 만료 처리를 구현하였습니다.
- RedisTemplate 활용
- 인증 코드 불일치와 인증 코드 만료를 구분하기 위해 VerificationCodeExpiredException를 추가하였습니다.

## 관련 이슈

- #242 

## 개발 유형

- [ ] 버그 수정
- [ ] 새로운 기능 개발
- [x] 리팩토링
- [ ] 테스트 코드 작성
- [ ] 문서 업데이트
- 기타 (아래에 설명 추가)

## 작업 기간

- 작업 시작일: (2024-04-16)
- 작업 종료일: (2024-04-16)

ghrltjdtprbs committed:
[refactor] code convention 적용

변경 사항 :
 - 공백, 줄바꿈 code convention 적용

ghrltjdtprbs committed:
[feat] 인증코드 유효시간 만료 exception 추가

문제 사항 :
 - 인증번호가 불일치하는 예외와 인증번호가 만료된 예외를 분리해야 할 필요성

변경 사항 :
 - VerificationCodeExpiredException 클래스 추가
 - ErrorCode 클래스에 EMAIL_VERIFICATION_CODE_EXPIRED 추가
 - redis에 사용자 이메일이 있으나, 인증번호가 일치하지 않을때는 기존 BAD_REQUEST를, 인증번호가 만료되어 redis에 해당 이메일이 없을 경우 GONE 예외를 발생시킵니다.

ghrltjdtprbs committed:
[refactor] Redis를 활용한 인증코드 만료기간 설정

문제 사항 :
 - 이메일로 전송된 인증코드의 만료기간을 설정할 필요성

변경 사항 :
 - 현재 prod서버가 내려가 테스트를 위해 LocalEmailService에도 유효기간 설정 로직을 추가하였습니다.
 - 유효기간 5분
 - RedisTemplate 사용

chimaek commented on src/main/java/com/fasttime/domain/member/exception/EmailTemplateLoadException.java:
```diff
@@ -11,5 +11,4 @@ public EmailTemplateLoadException() {
-
-}
\ No newline at end of file
+}
```
코드 리뷰:
1. 파일 끝에 새 줄이 없습니다. 커밋 시에 이것을 고려하십시오.
2. `EmailTemplateLoadException` 생성자에서 전달된 인수를 처리하지 않고 있으므로, 이 생성자가 필요한지 확인하세요.
3. 주석이 현재 없으므로, 코드의 목적과 해당 클래스의 역할을 설명하는 주석을 추가하는 것이 좋습니다.
4. 예외 생성자 내부의 `super(ERROR_CODE)`는 `ERROR_CODE`를 지원해야 합니다. 올바른 에러 코드 값을 전달하는지 확인하세요.

chimaek commented on src/main/java/com/fasttime/domain/member/exception/VerificationCodeExpiredException.java:
```diff
@@ -0,0 +1,14 @@
+
+        super(ERROR_CODE);
+    }
+}
```
이 코드는 새로운 패키지에 `VerificationCodeExpiredException` 클래스를 정의하고, `ApplicationException` 클래스를 확장하여 이메일 확인용 코드가 만료된 경우 해당 예외를 발생시킨다. 

잠재적인 버그나 개선점:
1. 생성자에 메시지를 추가하는 것이 도움이 될 수 있음.
2. 에러 코드 상수 사용시 재사용 가능한 모듈화 방식 고려.
3. 문서화 주석 추가.

개선 제안:
- 생성자를 수정해 에러 메시지 전달을 허용할 수 있도록 합니다.
- 사용되는 상수들을 외부 리소스에서 가져오도록하여 유지보수성 높일 것입니다.
- 클래스 및 메서드에 JavaDoc 주석을 추가하여 사용법 및 예외처리 방법을 명확히합니다.

chimaek commented on src/main/java/com/fasttime/domain/member/service/LocalEmailService.java:
```diff
@@ -57,4 +69,4 @@ private String generateAuthCode() {
     }
-}
\ No newline at end of file
+}
```
이 코드에 대한 간단한 코드 리뷰:

1. **개선 사항**:
   - `@Autowired`를 사용하여 `RedisTemplate`을 주입하고 있으나, 생성자 주입을 고려해볼 수 있습니다.
   - 코드 상단에 있는 import문은 정리될 필요가 있습니다.

2. **버그 위험**:
   - 인증 코드 저장 및 가져오기에 대한 로직이 변경되었는데, 이로 인해 캐싱 문제 등이 발생할 수 있으므로 테스트가 요구됩니다.

3. **주의 사항**:
   - Redis를 사용할 때의 성능과 안정성을 고려하여 Redis 클러스터 구성 및 예외 처리 전략이 필요합니다.

4. **일반 설명**:
   - 코드의 가독성과 유지 보수성을 높이기 위해 주석 및 메소드 이름을 명확하게 작성하는 것이 좋습니다.

5. **추가 고려 사항**:
   - Redis 연결 관리와 장애 처리를 강화하기 위해 RedisTemplate 설정을 개선할 수 있습니다.

전체적으로 Redis를 통한 인증 코드 저장으로 전환하는데 있어서 적절한 접근으로 보이지만, 성능 및 안정성에 대한 검증이 필요합니다.

chimaek commented on src/main/java/com/fasttime/domain/member/service/ProdEmailService.java:
```diff
@@ -29,29 +31,35 @@
+        return storedCode.equals(code);
     }
 
     private String generateAuthCode() {
```
코드 리뷰를 진행해보겠습니다.

1. **버그 위험:**
    - `verificationCodes` 맵을 통한 코드 저장 방식 대신, Redis를 사용하는 변경이 필요했습니다. 그러나 `verifyEmailCode` 메소드 내에서는 예외 처리가 누락되어 있습니다. 만일 Redis에 해당 이메일 주소에 대한 코드가 없을 경우 `null`을 반환하며, 여기서 `NullPointerException`이 발생할 수 있습니다.
   
2. **개선 제안:**
    - 비동기 작업을 고려하여 Java Mail 전송을 개선할 수 있습니다.
    - 인증 코드의 생성과 관리를 위한 명확한 인터페이스 및 인증 코드 관리 클래스 추가가 도움이 될 수 있습니다.
    - 리팩토링을 통해 `generateAuthCode` 메소드를 분리하여 코드 가독성을 향상시킬 수 있습니다.
     
3. **일반적인 리뷰:**
    - Lombok 어노테이션이 적절하게 사용되고 있습니다.
    - 거의 모든 것이 잘 구조화되어 있으며, Redis를 활용한 인증 코드 저장이 큰 개선입니다.

4. **추가 주의 사항:**
    - 코드 스니펫의 일부가 누란됐는지, 기존 코드와의 수정 비교가 더 도움이 된다면 반영도 가능합니다.

결론적으로, 보안 및 안정성을 위해 `NullPointerException`이 발생하지 않도록 애플리케이션을 보호하고 예외 처리를 완벽하게 구현하는 것이 관련된 주요 개선 사항입니다.

chimaek commented on src/main/java/com/fasttime/global/exception/ErrorCode.java:
```diff
@@ -9,6 +9,7 @@ public enum ErrorCode {
+    EMAIL_VERIFICATION_CODE_EXPIRED(HttpStatus.GONE, "인증 코드의 유효 기간이 만료되었습니다."),
 
     // MEMBER
     MEMBER_NOT_FOUND(HttpStatus.NOT_FOUND, "존재하지 않는 회원입니다."),
```
이 코드 패치의 요약은 다음과 같습니다:

1. `ErrorCode` 열거형에 새로운 상태 코드 `EMAIL_VERIFICATION_CODE_EXPIRED`가 추가되었습니다.
2. 새로운 상태 코드는 `HttpStatus.GONE`을 사용하며, "인증 코드의 유효 기간이 만료되었습니다."라는 설명이 포함되어 있습니다.

개선 제안:
- 새로운 에러 코드에 대한 테스트 케이스를 작성하여 적절한 동작을 확인할 수 있도록 합니다.
- 에러 메시지에 변수를 동적으로 추가하는 기능을 고려해 볼 수 있습니다.

ghrltjdtprbs committed:
[feat] REST Docs에 인증코드 유효시간 설명 추가

문제 사항 :
 - 프론트에게 전송받은 인증코드 유효시간을 알릴 필요성

변경 사항 :
 - email-api.adoc 파일에 유효시간에 대한 설명을 추가하였습니다.

chimaek commented on src/docs/asciidoc/email/email-api.adoc:
```diff
@@ -30,6 +30,8 @@ include::{snippets}/email-send-verification/response-fields.adoc[]
+
 
 === HttpRequest
 
```
이 코드 패치에 대해 간단한 코드 리뷰를 제공하겠습니다.

1. **잠재적 버그 위험:**
   - 인증 코드의 유효시간이 5분으로 설정되었지만, 해당 시간에 대한 구현이 보이지 않으므로 동작하는 방식을 확인해야 합니다.
  
2. **개선 제안:**
   - 유효시간 내에 인증 코드가 만료되는 경우를 처리하는 방법이 필요합니다. 만료되면 적절한 오류 처리나 메시지를 전송해야 합니다.
   - 인증 코드 유효성 검사 시, 사용자가 재시도할 수 있는 방법을 제공하는 것이 좋습니다.

이러한 점들을 확인하고 수정한다면 코드의 안전성과 사용자 경험이 향상될 것입니다.
